add a system variable tiflash_replica_read (#14179)
pingcap/docs-cn: system-variables.md

    * 大于 0: 表示使用细粒度 shuffle 功能，下推到 TiFlash 的窗口函数会以多线程方式执行，并发度为： min(`tiflash_fine_grained_shuffle_stream_count`, TiFlash 节点物理线程数)
- 理论上窗口函数的性能会随着该值的增加线性提升。但是如果设置的值超过实际的物理线程数，反而会导致性能下降。

> 

Reviewer: 关于 TiFlash 的副本选择策略是否需要在 https://docs.pingcap.com/zh/tidb/dev/use-tidb-to-read-tiflash)](https://docs.pingcap.com/zh/tidb/dev/use-tidb-to-read-tiflash 文档里也简单说明下呢，这样更方便用户知道这个功能

Reviewer 2: 我建议在 https://docs.pingcap.com/zh/tidb/dev/create-tiflash-replicas#%E8%AE%BE%E7%BD%AE%E5%8F%AF%E7%94%A8%E5%8C%BA 这里增加一部分说明。上面的 link 文档主要是 tikv 和 tiflash 之间的选择，不太适合。

### `tiflash_replica_read` <span class="version-mark">从 v7.2.0 版本开始引入</span>

> 

Author: 是一个事情，就是 tiflash 副本放在哪里，用户的相关设置方式要做全套

> 

Author: 这里关注那个区域属性的文档就行

- 作用范围：SESSION | GLOBAL
- 持久化至集群：是
- 类型：枚举
- 默认值："all_replicas"
- 可选值："all_replicas"、"closest_adaptive"、"closest_replicas"

> - 该变量用于设置当查询需要使用 TiFlash 引擎时，TiFlash 节点的选择策略。

Reviewer 2: ```suggestion
- 该变量用于设置当查询需要使用 TiFlash 引擎时，TiFlash 副本的选择策略。
```

>   - "all_replicas" 表示使用所有可用节点进行分析计算。

Reviewer 2: ```suggestion
  - "all_replicas" 表示使用所有的 TiFlash 副本进行分析计算。
```

>   - "closest_adaptive" 表示使用与入口 TiDB 相同区域的节点。如果无法访问所有的 TiFlash 数据，则查询将通过与入口 TiDB 相同区域的节点访问来自其他区域的 TiFlash 节点。

Reviewer 2: ```suggestion
  - "closest_adaptive" 表示尽量使用与入口 TiDB 相同区域的 TiFlash 副本进行分析计算。如果无法访问所有的 TiFlash 数据，则查询将使用与入口 TiDB 不同区域的 TiFlash 副本及对应的 TiFlash 节点。
```

>   - "closest_replicas" 表示仅使用与入口 TiDB 相同区域的节点。如果无法访问所有的 TiFlash 数据，则查询将报错。

Reviewer 2: ```suggestion
  - "closest_replicas" 表示仅使用与入口 TiDB 相同区域的 TiFlash 副本。如果无法访问所有的 TiFlash 数据，则查询将报错。
```

- 特殊情况

>   - 如果 TiDB 节点未设置区域属性，并且 TiFlash 节点选择策略不是 "all_replicas"，则会忽略 TiFlash 节点选择策略，将使用所有 TiFlash 节点进行 TiFlash 查询。并且会有一个警告消息：The variable tiflash_replica_read is ignored。

Reviewer 2: ```suggestion
  - 如果 TiDB 节点未设置区域属性，并且 TiFlash 副本选择策略不是 "all_replicas"，则会忽略 TiFlash 副本选择策略，将使用所有 TiFlash 副本进行 TiFlash 查询。并且会有一个警告消息：The variable tiflash_replica_read is ignored。
```

  - 如果 TiFlash 节点未设置区域属性，则将其视为不属于任何区域的节点。

> 

Reviewer: 顺便问下，这个区域属性是通过 https://docs.pingcap.com/zh/tidb/dev/schedule-replicas-by-topology-labels 设置的吗

> - 特殊情况
>   - 如果 TiDB 节点未设置区域属性，并且 TiFlash 副本选择策略不是 `all_replicas` 时，TiFlash 引擎将忽略 TiFlash 副本选择策略，使用所有 TiFlash 副本进行 TiFlash 查询，并且返回警告 `The variable tiflash_replica_read is ignored`。
>   - 如果 TiFlash 节点未设置区域属性，则将其视为不属于任何区域的节点。

Reviewer: ```suggestion

> **注意：**
> 
 > - 如果 TiDB 节点未设置[区域属性](/schedule-replicas-by-topology-labels.md#设置-tidb-的-labels可选)，并且 TiFlash 副本选择策略不是 `all_replicas` 时，TiFlash 引擎将忽略 TiFlash 副本选择策略，使用所有 TiFlash 副本进行 TiFlash 查询，并且返回警告 `The variable tiflash_replica_read is ignored`。
 > - 如果 TiFlash 节点未设置[区域属性](/schedule-replicas-by-topology-labels.md#设置-tikv-和-tiflash-的-labels)，则将其视为不属于任何区域的节点。
```

    * 大于 0: 表示使用细粒度 shuffle 功能，下推到 TiFlash 的窗口函数会以多线程方式执行，并发度为： min(`tiflash_fine_grained_shuffle_stream_count`, TiFlash 节点物理线程数)
- 理论上窗口函数的性能会随着该值的增加线性提升。但是如果设置的值超过实际的物理线程数，反而会导致性能下降。

> ### `tiflash_replica_read` <span class="version-mark">从 v7.2.0 版本开始引入</span>

Reviewer: ```suggestion
### `tiflash_replica_read` <span class="version-mark">从 v7.3.0 版本开始引入</span>
```

- 作用范围：SESSION | GLOBAL
- 持久化至集群：是
- 类型：枚举型
- 默认值：`all_replicas`
- 可选值：`all_replicas`、`closest_adaptive`、`closest_replicas`
- 该变量用于设置当查询需要使用 TiFlash 引擎时，TiFlash 副本的选择策略。

>   - `all_replicas` 表示使用所有的 TiFlash 副本进行分析计算。
>   - `closest_adaptive` 表示尽量使用与当前发起查询请求的 TiDB 节点相同区域的 TiFlash 副本进行分析计算。如果此区域的 TiFlash 副本未包含查询所需的全部数据，则再使用其他区域的 TiFlash 副本及对应的 TiFlash 节点。
>   - `closest_replicas` 表示仅使用与发起当前查询请求的 TiDB 节点相同区域的 TiFlash 副本进行分析计算。如果此区域的 TiFlash 副本未包含查询所需的全部数据，则查询将报错。

Reviewer: ```suggestion
    - `all_replicas` 表示使用所有的 TiFlash 副本进行分析计算。
    - `closest_adaptive` 表示尽量使用与当前发起查询请求的 TiDB 节点相同区域的 TiFlash 副本进行分析计算。如果此区域的 TiFlash 副本未包含查询所需的全部数据，则再使用其他区域的 TiFlash 副本及对应的 TiFlash 节点。
    - `closest_replicas` 表示仅使用与发起当前查询请求的 TiDB 节点相同区域的 TiFlash 副本进行分析计算。如果此区域的 TiFlash 副本未包含查询所需的全部数据，则查询将报错。
```

> **注意：**
> 

>  > - 如果 TiDB 节点未设置[区域属性](/schedule-replicas-by-topology-labels.md#设置-tidb-的-labels可选)，并且 TiFlash 副本选择策略不是 `all_replicas` 时，TiFlash 引擎将忽略 TiFlash 副本选择策略，使用所有 TiFlash 副本进行 TiFlash 查询，并且返回警告 `The variable tiflash_replica_read is ignored`。

Reviewer 3: ```suggestion
> - 如果 TiDB 节点未设置[区域属性](/schedule-replicas-by-topology-labels.md#设置-tidb-的-labels可选)，并且 TiFlash 副本选择策略不是 `all_replicas` 时，TiFlash 引擎将忽略 TiFlash 副本选择策略，使用所有 TiFlash 副本进行 TiFlash 查询，并且返回警告 `The variable tiflash_replica_read is ignored`。
```

>  > - 如果 TiFlash 节点未设置[区域属性](/schedule-replicas-by-topology-labels.md#设置-tikv-和-tiflash-的-labels)，则将其视为不属于任何区域的节点。

Reviewer 3: ```suggestion
> - 如果 TiFlash 节点未设置[区域属性](/schedule-replicas-by-topology-labels.md#设置-tikv-和-tiflash-的-labels)，则将其视为不属于任何区域的节点。
```

- 类型：布尔型
- 默认值：`ON`
- 这个变量用于控制计算窗口函数时是否采用高精度模式。

> ### tidb_tiflash_node_selection_policy

Reviewer: system-variables.md 里的变量是按照字母顺序排列的，请调整下它的位置
